keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlVerificationRepository.kt

```diff
@@ -162,6 +168,77 @@ class SqlVerificationRepository(
       }
     }
 
+  /**
+   * Query the repository for the states of multiple contexts.
+   *
+   * This call is semantically equivalent to
+   *    contexts.map { context -> this.getStates(context) }
+   *
+   * However, it's implemented as a single query for efficiency.
+   *
+   * @param contexts a list of verification contexts to query for state
+   * @return a list of maps of verification ids to states, in the same order as the contexts. If there are no
+   *         verification states associated with a context, the resulting map will be empty.
+   */
+  override fun getStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>> {
+    /**
+     * In-memory database table representation of the set of contexts we want to query for
+     *
+     * Columns:
+     *   ind - index that encodes the original list order, to ensure results are in same order
+     *   environment_name
+     *   artifact_reference
+     *   artifact_version
+     */
+    val contextTable = ContextTable(contexts, jooq)
+
+    /**
+     * This function guarantees that the number of output elements match the number of input elements.
+     * So we use left joins on the givenVersions table
+     */
+    return jooq.select(
+
+      contextTable.IND,
+      VERIFICATION_STATE.VERIFICATION_ID,
+      VERIFICATION_STATE.STATUS,
+      VERIFICATION_STATE.STARTED_AT,
+      VERIFICATION_STATE.ENDED_AT,
+      VERIFICATION_STATE.METADATA)
+      .from(contextTable.table)
+      .leftJoin(ENVIRONMENT)
+      .on(ENVIRONMENT.NAME.eq(contextTable.ENVIRONMENT_NAME))
+      .leftJoin(DELIVERY_CONFIG)
+      .on(DELIVERY_CONFIG.UID.eq(ENVIRONMENT.DELIVERY_CONFIG_UID))
+      .leftJoin(DELIVERY_ARTIFACT)
+      .on(DELIVERY_ARTIFACT.REFERENCE.eq(contextTable.ARTIFACT_REFERENCE))
+      .leftJoin(VERIFICATION_STATE)
+      .on(VERIFICATION_STATE.ARTIFACT_UID.eq(DELIVERY_ARTIFACT.UID))
+      .and(VERIFICATION_STATE.ARTIFACT_VERSION.eq(contextTable.ARTIFACT_VERSION))
+      .and(DELIVERY_ARTIFACT.DELIVERY_CONFIG_NAME.eq(DELIVERY_CONFIG.NAME))
+      .and(VERIFICATION_STATE.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+
+      // execute the query
+      .fetch()
+
+      // sort the results by the "ind" (index) column, so that outputs are same order as inputs
+      .groupBy { (index, _, _, _, _, _) -> index as Long }
+      .toSortedMap()
+      .values
+
+      // convert List<Record> to Map<String,VerificationState>, where the string is the verification id
+      .map { records ->
+        records
+          // since we do a left join, there may be rows where there is no corresponding records in the
+          // verification_state database, so we filter them out, which will result in an empty map
```

Comment:
> which will result in an empty map

Should we additionally filter out those empty maps from the list, then?

Reply:
No, because the input list has to be the exact same size as the output list so we can match up VerificationContext objects with VerificationState objects. (See [this earlier comment](https://github.com/spinnaker/keel/pull/1752#discussion_r564813023)).